### PR TITLE
Refactor application_base for teacher applications

### PR DESCRIPTION
**Pull request by megcrenshaw**
This sets up application_base to account for what's needed for teacher applications to save.

Teacher applications will have a "status: incomplete" added to `form_data` when they save. So status needs to be set on that if possible. Also updated the form.rb file to ignore validation on an incomplete status.

Also did some refactors to group callbacks together.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/application_base.rb:114]: Moved to line 60, before_validation. This is probably best to happen before_create, so happy to change it back––was trying to reduce the number of different callbacks we had

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/application_base.rb:111]: These are all only happening before_validation now

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/application_base.rb:104]: I don't have to do this, but having 20 lines for the same thing was bothering me

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/application_base.rb:225]: moved up to be with the callbacks ... it was lonely

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/application_base.rb:253]: Moved these up to the top of the file as well. Tbh, I totally missed the status validation for a while, not expecting it to be down here ...

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [inline dashboard/test/models/pd/application/application_base_test.rb:33]: Since the status is set on validation now, we have to actually create a whole application. Not sure how to avoid using TeacherApplication or another application that uses ApplicationBase

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [inline dashboard/test/models/pd/application/application_base_test.rb]: I originally used `application = create TEACHER_APPLICATION_FACTORY` instead, without a save. Wanted to stick to using ApplicationBase. I have no idea why this way works––status is being set `before_validation`.

I don't know why other validations aren't failing ("Form data is required, Type is required, User is required, Application type is not included in the list, Application year is not included in the list"). According to https://guides.rubyonrails.org/getting_started.html#validations-and-displaying-error-messages, validations happen before the object is saved. Why would the `before_validation` be triggered but not the other validations?

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [inline dashboard/app/models/pd/application/application_base.rb:63]: what value is this `if` clause adding? I get the desire to not change more than necessary, but it seems like this is saying "if create or update", which seems like it might cover all of the cases where a validation would be run. 

I would lean on test coverage to answer this -- once you have tests covering the status field for all the codepaths you care about, I would suggest trying to remove this if clause to simplify things a bit. if nothing breaks then you don't need it

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [inline dashboard/app/models/pd/application/application_base.rb:62]: when is form data getting set? I think it could help to write some code comments documenting your strategy here:
1. set form_data (when?)
2. populate other fields from form_data before_validation, presumably because it happens on both `save` and `update`

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [reply dashboard/app/models/pd/application/application_base.rb:114]: I see what you're aiming for, but this one makes more sense to me to do before_create -- because this is something that will only ever be set on create, never on update.

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [inline dashboard/app/models/pd/application/application_base.rb:74]: just confirming my understanding that this is ok to leave as `before_save`, because a newly created application is never in the accepted state, it is (usually? always?) unreviewed or incomplete.

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [reply dashboard/app/models/pd/application/application_base.rb:104]: these boilerplate definitions are annoying but useful, since they make it possible for IDEs to jump to the definition of these methods. that said there are plenty of places that rails auto-generates methods (especially ones ending in `?`) in ways that are hard for an IDE to locate. I'm fine to go with your  personal preference here (and @tim-dot-org's if they have one?) since y'all are spending the most time in this code.

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (davidsbailey) [reply dashboard/test/models/pd/application/application_base_test.rb]: when using `save`, failing validations cause `nil` to be returned, so that could be what's happening here. try switching to `save!`, which will raise an error saying which validation failed.

**Review by davidsbailey — COMMENTED** [review]
Awesome work @megcrenshaw ! I love the direction this is heading in.

one concern with refactoring PRs is that you are relying on existing test coverage. in this case that includes other application types such as principle application and facilitator application. I would suggest looking to see if there is at least one UI test covering each of those. @tim-dot-org  any chance you have a sense of the level of test coverage there? I see UI tests for "principal approval" but not principal application.

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:104]: Yeah, I've run into the problem of not being able to find things in IDEs ... @tim-dot-org would love to hear what you think. Would be happy to revert

Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:104]: I could also add a comment that says "generates accepted?, incomplete?, pending?, unreviewed?, waitlisted? methods" but then we get into a potential problem of the comment being out of date

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/pd/application/application_base_test.rb]: Ah, thank you. Well, it's nice to avoid the validations causing a failure in this case tehe. Or we could create using `TEACHER_APPLICATION_FACTORY` and not need a save (was trying to avoid that to stay inside of ApplicationBase) ... do you have a strong preference?

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:63]: It's possible a status gets updated manually, without the form data changing (e.g. when an RP changes the status). We don't want the status to be overwritten to the form data in that case.

Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:63]: As for tests, there was already a test for the default status being unreviewed. I added a test "form data overwrites default status" ... so also add a test that we can manually update the form's status? I just added some refactors to other tests that were failing because of this problem––thoughts?

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:74]: Yes, exactly––would a comment be helpful here as well?

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (davidsbailey) [reply dashboard/test/models/pd/application/application_base_test.rb]: I am not sure I understand the situation. why did you need to add a save (or anything else) here?

if the situation is that this test is suddenly failing, i would say that, it doesn't really make sense to me to be instantiating the ApplicationBase class here, because that's not something we ever need to do in our application -- we are only every instantiating children of this class (please correct me if I'm wrong there). based on that, here are my thoughts:
1.  instantiate teacher application here instead
2.  should other classes which extend ApplicationBase should also have similar tests?
3.  should these tests move into the inheriting class's test? e.g. teacher_application_test.rb.

based on the pattern existing pattern here, I'd suggest doing (1), not doing (3), and poking around a bit to see if (2) seems necessary based on the changes you are making.

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [reply dashboard/app/models/pd/application/application_base.rb:63]: sounds good to me on both counts. thanks!

File: dashboard/app/models/pd/application/application_base.rb
Comment (davidsbailey) [reply dashboard/app/models/pd/application/application_base.rb:74]: yes, I think that would be helpful to comment, thanks!

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:114]: Good call––moved it back

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/pd/application/application_base_test.rb]: Changed it back to (1). As for (2):
- `teacher_application_test.rb` has a test making sure we can go from unreviewed to pending
- there are tests that shouldn't work in both `facilitator1819_application_test.rb` and `facilitator1819_application_test.rb` but they are working for different reasons (mostly involving the "locked" state of the application being tested correctly and overriding whatever's happening with the "status")
- `principal_approval_application_test.rb` has a test making sure status can be nil, but the test feels a bit wonky because the status is never actually nil (e.g. in staging, the status is set as unreviewed upon `build` even though we say `status: nil` in the test)

So, options from here:
1) I refactor tests from all these files before shipping out this PR, and add tests explicitly making sure the functionality is consistent from before
2) I add a separate ticket to refactor these (will it ever get picked up?), as the tests don't warrant a refactor based on the changes I'm making, but rather on how we're currently testing 
I *think* (1) would take about half a day––I'd mostly manually update the status after creating the application.

File: dashboard/app/models/pd/application/application_base.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/application_base.rb:62]: FWIW, I changed `:set_type_and_year back` to happening `after_initialize`: there's no reason it needs to happen before every validation

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (davidsbailey) [reply dashboard/test/models/pd/application/application_base_test.rb]: As for (2), I meant to ask if, by moving a test off of ApplicationBase onto TeacherApplication, it left a gap in test coverage on those other models (facilitator and principal applications), specifically around verifying the default status after creating (or initializing) an object. i.e., should we add a "default status is unreviewed" test case for those other models.

I can take a look this afternoon at the possible test refactors you are talking about, but I wouldn't let those block this PR.

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (davidsbailey) [reply dashboard/test/models/pd/application/application_base_test.rb:33]: what do you mean by a "whole application" -- do you just mean that you're instantiating the teacher application (a subclass), or do you mean to emphasize that in order to do that you have to fill in all the fields so that it passes validations? I think these are both fine -- the factory seems to take care of filling in those details for you, and I honestly don't see the value in a unit tests which just instantiates and abstract base class, something we'd never do in our application. if there was some specific logic within ApplicationBase that we wanted to test, we could extract that logic into a particular method on ApplicationBase and write a unit test specifically against that method.

**Comment by davidsbailey** [comment]
thanks for your patience Meg, a lot of this code is new to me so you're getting some questions that are part figuring it out and part (somewhat unfiltered) questions I would ask if I was trying to make a change in this space. I can take another closer look this afternoon.

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/pd/application/application_base_test.rb:33]: Yes, all sounds good

Your guess was right: "whole application," aka one that passes validations

File: dashboard/test/models/pd/application/application_base_test.rb
Comment (megcrenshaw) [reply dashboard/test/models/pd/application/application_base_test.rb]: It is unclear for me what the default status should be for the principal application. The tests indicate "nil" status by default (and we have a test in application_base for that case), but it looks like application_base writes it as "unreviewed" (in staging). FacilitatorApplication seems clear that it is "unreviewed" by default, although there isn't a strong test for it right now.

I don't even think we do facilitator applications any more. It's unclear to me how much time to invest in making sure those work––not sure best practice here

Comment (davidsbailey) [reply dashboard/test/models/pd/application/application_base_test.rb]: ok, thanks for looking, I think it is fine to just change the unit test to create a teacher application.

**Comment by megcrenshaw** [comment]
> thanks for your patience Meg, a lot of this code is new to me so you're getting some questions that are part figuring it out and part (somewhat unfiltered) questions I would ask if I was trying to make a change in this space. I can take another closer look this afternoon.

@davidsbailey Your comments are very helpful. As for UI stuff, I don't see existing UI coverage for facilitator and principal applications. (There is `PrincipalApprovalTest.js` but I'm not sure that's the coverage you're asking about.) @tim-dot-org 's tests for teacher applications are great, and I added to them with new functionality added (see https://github.com/code-dot-org/code-dot-org/pull/44441)

**Comment by davidsbailey** [comment]
@megcrenshaw for other application types, could you please try Tim's steps in [slack](https://codedotorg.slack.com/archives/C02EEGWLHR8/p1643322115176200) for testing just the principal application locally? the goal would just be to make sure that there is a not a regression in getting it to submit -- I wouldn't wouldn't worry about any of the later steps like accepting. hopefully, after the refactoring in this PR is done, you won't be messing with ApplicationBase so much and won't need to manually test this again in subsequent PRs.

File: dashboard/test/factories/pd_factories.rb
Comment (davidsbailey) [inline dashboard/test/factories/pd_factories.rb:814]: nice addition and use of this hash factory !

**Review by davidsbailey — APPROVED** [review]
LGTM after manually testing the principal application locally. Thanks for your patience.

**Comment by megcrenshaw** [comment]
> @megcrenshaw for other application types, could you please try Tim's steps in [slack](https://codedotorg.slack.com/archives/C02EEGWLHR8/p1643322115176200) for testing just the principal application locally? the goal would just be to make sure that there is a not a regression in getting it to submit -- I wouldn't wouldn't worry about any of the later steps like accepting. hopefully, after the refactoring in this PR is done, you won't be messing with ApplicationBase so much and won't need to manually test this again in subsequent PRs.

Sounds good –– tested locally, and things look good (note the status as "unreviewed" which matches what's in production as well)
Screen Shot 2022-01-28 at 10 01 09 AM
Screen Shot 2022-01-28 at 10 41 07 AM
Screen Shot 2022-01-28 at 10 42 06 AM

File: dashboard/app/models/pd/application/application_base.rb
Comment (tim-dot-org) [reply dashboard/app/models/pd/application/application_base.rb:104]: yeah, i think the comment would be a good solution! that list has to be manually updated anyways, if the comment is right there it'll be easy to update alongside

**Review by tim-dot-org — APPROVED** [review]
sorry i missed the pings! This looks great, thanks for putting things in more reasonable places in application base!

**Pull request merged**
